internal/clients/compute/group/group.go

```diff
@@ -0,0 +1,419 @@
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/internal/clients/compute"
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/internal/utils"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// APIClient is a wrapper around IONOS Service
+type APIClient struct {
+	*clients.IonosServices
+}
+
+// Client is a wrapper around IONOS Service Group methods
+type Client interface {
+	CheckDuplicateGroup(ctx context.Context, groupName string) (*sdkgo.Group, error)
+	GetGroupID(group *sdkgo.Group) (string, error)
+	GetGroup(ctx context.Context, groupID string) (sdkgo.Group, *sdkgo.APIResponse, error)
+	GetGroupMembers(ctx context.Context, groupID string) ([]string, *sdkgo.APIResponse, error)
+	GetGroupResourceShares(ctx context.Context, groupID string) ([]v1alpha1.ResourceShare, *sdkgo.APIResponse, error)
+	CreateGroup(ctx context.Context, group sdkgo.Group) (sdkgo.Group, *sdkgo.APIResponse, error)
+	UpdateGroup(ctx context.Context, groupID string, group sdkgo.Group) (sdkgo.Group, *sdkgo.APIResponse, error)
+	AddGroupMember(ctx context.Context, groupID, userID string) (*sdkgo.APIResponse, error)
+	RemoveGroupMember(ctx context.Context, groupID, userID string) (*sdkgo.APIResponse, error)
+	UpdateGroupMembers(ctx context.Context, groupID string, membersIn MembersUpdateOp) error
+	AddResourceShare(ctx context.Context, groupID string, share v1alpha1.ResourceShare) (*sdkgo.APIResponse, error)
+	RemoveResourceShare(ctx context.Context, groupID string, share v1alpha1.ResourceShare) (*sdkgo.APIResponse, error)
+	UpdateResourceShare(ctx context.Context, groupID string, share v1alpha1.ResourceShare) (*sdkgo.APIResponse, error)
+	UpdateGroupResourceShares(ctx context.Context, groupID string, sharesIn SharesUpdateOp) error
+	DeleteGroup(ctx context.Context, groupID string) (*sdkgo.APIResponse, error)
+	GetAPIClient() *sdkgo.APIClient
+}
+
+// MembersUpdateOp groups memberIDs in sets depending on the operation in which they will be used
+type MembersUpdateOp struct {
+	Add, Remove sets.Set[string]
+}
+
+// SharesUpdateOp groups resource shares in sets depending on the operation in which they will be used
+type SharesUpdateOp struct {
+	Add, Update, Remove sets.Set[v1alpha1.ResourceShare]
+}
+
+// CheckDuplicateGroup based on groupName
+func (cp *APIClient) CheckDuplicateGroup(ctx context.Context, groupName string) (*sdkgo.Group, error) {
+	groups, _, err := cp.ComputeClient.UserManagementApi.UmGroupsGet(ctx).Depth(utils.DepthQueryParam).Execute()
+	if err != nil {
+		return nil, err
+	}
+	matchedItems := make([]sdkgo.Group, 0)
+
+	if groups.Items != nil {
+		for _, item := range *groups.Items {
+			if item.Properties != nil && item.Properties.Name != nil && *item.Properties.Name == groupName {
+				matchedItems = append(matchedItems, item)
+			}
+		}
+	}
+
+	if len(matchedItems) == 0 {
+		return nil, nil
+	}
+	if len(matchedItems) > 1 {
+		return nil, fmt.Errorf("error: found multiple groups with the name %v", groupName)
+	}
+	return &matchedItems[0], nil
```

Comment:
I am not sure what will happen in this case when the GC runs. my understanding is you return the address of an element in a slice. when that slice is GC(garbage collected) then that address points to an invalid memory. I am not sure if the GC will be able to identify this case and not freeing the slice.

Reply:
Even though the `matchedItems` slice variable goes out of scope after the return, the underlying array of this slice will still be kept and not marked for GC as long as there are references to it.

There is actually a different issue to this style, which is that the entire underlying array for `matchedItems` is being kept in memory while only 1 element of it can still be accessed, so value return should be used.

Overall this function is better off just returning the `ID` string of the duplicate group, if any:
CheckDuplicateGroup(ctx context.Context, groupName string) (groupID, error)

Reply:
fixed ([df32e25](https://github.com/ionos-cloud/crossplane-provider-ionoscloud/pull/140/commits/df32e25c674273b256ff319f4d070564728cbeeb))